diff --git a/src/xrt/targets/common/target_instance.c b/src/xrt/targets/common/target_instance.c
index e10ba8358..ad10b41fa 100644
--- a/src/xrt/targets/common/target_instance.c
+++ b/src/xrt/targets/common/target_instance.c
@@ -27,6 +27,7 @@
 #include "target_instance_parts.h"
 
 #include <assert.h>
+#include <unistd.h> // for sleep()
 
 #ifdef XRT_OS_ANDROID
 #include "android/android_instance_base.h"
@@ -77,6 +78,7 @@ t_instance_create_system(struct xrt_instance *xinst,
        usys = u_system_create();
        assert(usys != NULL); // Should never fail.
 
+       os_nanosleep(U_TIME_1MS_IN_NS * 10000);
        xret = u_system_devices_create_from_prober( //
            xinst,                                  // xinst
            &usys->broadcast,                       // broadcast
@@ -162,6 +164,7 @@ xrt_instance_create(struct xrt_instance_info *ii, struct xrt_instance **out_xins
 
        XRT_TRACE_MARKER();
 
+       os_nanosleep(U_TIME_1MS_IN_NS * 10000);
        int ret = xrt_prober_create_with_lists(&xp, &target_lists);
        if (ret < 0) {
                return XRT_ERROR_PROBER_CREATION_FAILED;
